parser/Parser.cc

```diff
@@ -16,72 +16,122 @@ extern const char *dclassStrings[];
-                e.replaceWith("Replace with `it`", loc, "it");
-                break;
-            case ruby_parser::dclass::MissingCommaBetweenKwargs:
-                e.replaceWith("Insert a comma", loc, ", ");
-                break;
-            case ruby_parser::dclass::CurlyBracesAroundBlockPass: {
-                auto source = loc.source(gs);
-                if (source.has_value()) {
-                    auto replacement = string(source.value());
-                    size_t lCurlyPos = replacement.find("{");
-                    size_t rCurlyPos = replacement.rfind("}");
-                    if (lCurlyPos != string::npos && rCurlyPos != string::npos) {
-                        replacement[lCurlyPos] = '(';
-                        replacement[rCurlyPos] = ')';
-                        e.replaceWith("Replace the curly braces with parens", loc, replacement);
-                    }
+core::Loc rangeToLoc(const core::GlobalState &gs, core::FileRef file, const ruby_parser::diagnostic::range &range) {
+    uint32_t maxOff = file.data(gs).source().size();
+    core::Loc loc(file, translatePos(range.beginPos, maxOff - 1), translatePos(range.endPos, maxOff));
+    return core::Loc(file, translatePos(range.beginPos, maxOff - 1), translatePos(range.endPos, maxOff));
+}
+
+core::ErrorClass dclassToErrorClass(ruby_parser::dclass diagClass) {
+    switch (diagClass) {
+        case ruby_parser::dclass::DedentedEnd:
+            return core::errors::Parser::ErrorRecoveryHint;
+        default:
+            return core::errors::Parser::ParserError;
+    }
+}
+
+void explainError(core::GlobalState &gs, core::FileRef file, core::ErrorBuilder &e, core::Loc loc,
+                  ruby_parser::diagnostic diag) {
+    switch (diag.error_class()) {
+        case ruby_parser::dclass::IfInsteadOfItForTest:
+            e.replaceWith("Replace with `it`", loc, "it");
+            break;
+        case ruby_parser::dclass::MissingCommaBetweenKwargs:
+            e.replaceWith("Insert a comma", loc, ", ");
+            break;
+        case ruby_parser::dclass::CurlyBracesAroundBlockPass: {
+            auto source = loc.source(gs);
+            if (source.has_value()) {
+                auto replacement = string(source.value());
+                size_t lCurlyPos = replacement.find("{");
+                size_t rCurlyPos = replacement.rfind("}");
+                if (lCurlyPos != string::npos && rCurlyPos != string::npos) {
+                    replacement[lCurlyPos] = '(';
+                    replacement[rCurlyPos] = ')';
+                    e.replaceWith("Replace the curly braces with parens", loc, replacement);
                 }
-                break;
             }
-            default:
-                break;
+            break;
         }
+        case ruby_parser::dclass::DedentedEnd:
+            ENFORCE(diag.extra_location().has_value());
+            e.addErrorLine(rangeToLoc(gs, file, diag.extra_location().value()), "Matching token was here");
```

Comment:
To be clear: we don't have any tests that check we point at the matching token?

Reply:
i'll add one